Another api

francoisthire opened the pull request:
We completely rework the API to make it proper and in a way to have a kernel safer. We also changed the build system to `dune` without the huge modifications of #194 . Warnings are the same as `ocamlc`.
We also remove the feature that each folder gives rise a module, but I think this is something we want on the long term, but it would imply more changes and the diff is getting huge now. We have updated the `Makefile` in consequences which I think makes it simpler. The advantage of using `dune` is also a better way to handle documentation.

# Kernel

## Basic

[x] Remove the `add_path` and `get_path` function from `kernel/basic.ml` to `api/dep.ml`
[x] Add modules `MidentSet` `NameSet` from `Dep`
## Rules

[x] Move functions `check_arity` and `check_linearity` from `Env` to `Rule`.
[x] Move exception constructors `NotEnoughArguments` and `NonLinearRule` from `Env` to `Rule` and add a location information

## Signature

[x] Add an exception `CannotExportModule`
[x] Make `Signature` independent from the file system. The `import` functions and `export` functions are modified consequently

# Parser

## Parser

The parser now communicates only with the kernel.

[x] The folder `parser`does not depend anymore on `Env`
[x] Introducing a type `Parser.t` which encapsulate the `input` where it is a string, a file or stdin
[x] Introducing an exception type `Parser.Parse_error`
[x] Removing functors `Parser_channel` and `Parse_string` in `Parser`. This is replace by functions `Parser.input_from_string` and `Parser.input_from_channel`
[x] Add a function `Parser.input_from_stdin`
[x] Add a function `Parser.md_of_input`
[x] Add a function `Parser.file_of_input`
[x] Add a function `Parser.close`

## Lexer
[x] Introducing an exception type `Lexer.lexer_error`

# api

## Env

[x] Remove `Env.S` and the functor `Env.Make`. This can be recover from the new interface and first-class modules
[x] Add a type `Env.t` which make the interface `Env` purer
[x] Add methods `Env.set_reduction_engine` and `Env.get_reduction_engine`
[x] Add method `Env.get_printer`
[x] Remove the type `Env.EnvError`. This simplifies the code a lot. Errors can be handled by `Processor`

## Entry

[x] Add en exception `Assert_error` (which replaces the one in `Env`)

## Dep

[x] rename `compute_ideps` from `compute_all_deps`
[x] New function from `Basic` : `add_path` and `get_path`
[x] New function `find_object_file`

## Pp

[x] Add a boolean `print_module_name` that always print the module name

## Processor

[x] Add `handle_processor` which applies a processor to the current environment. This method may raise an Env_error
[x] Add `handle_input` which process directly and `input`. One can put `hooks` which will be processed before and after reading the input
[x] Add `handle_files` which process a list of files. As `handle_input`, `hooks` can be branched.

## Errors

[x] Internally, we rework the errors code. (this should be changed see #193[)
[ ] Catch sys_error
[ ] Remove the backtrace

GuillaumeGen commented:
Looks great to me. Two suggested improvements:
 1 - Modify the comments in env.mli to take accountof the argument you added to most of the functions.
 2 - In processor.ml, the alias D for Debug is used twice, whereas there are four debug printings, so I would suppress this alias.

GuillaumeGen commented:
Il est recommandé de faire `make` avant de `push` (voire avant de commiter).

Gaspi commented:
> Rework the API so that it is `purer`.
> 
>     * Add a type `t` in module `Env`
>     * An environment has to be created first before parsing a new file
> 

This is a great !

> 
>     * Remove most of the functors of the API
> 
I think this deserve discussion. I'm not sure this is an improvement of the API because the new implementation relies on first-class modules instead.

In my opinion, first-class modules are a rather advanced feature of OCaml that should be avoided if possible. There wasn't a single occurrence of first-class modules before and now I see around 20-30 of them in the code of the API including several in the `parser` and `command` folders which are supposed to be understandable (if not trustable) by the user.

> Also, removing functors allows a better flexibility in practice because the syntax of first-class modules can be annoying.
> 
It seems to me that we should rather remove first-class modules then...

> This interface remove the usage of first-class modules except for the printer.
> 
People could understand the code base without even knowing about first-class modules before, this PR makes sure it's not the case anymore...

Do you think it's possible to split this PR in two: keep the new pure interface and postpone the use of first-class module ?

Gaspi commented:
Note that in the official documentation of OCaml  ([caml.inria.fr/pub/docs/manual-ocaml/index.html](caml.inria.fr/pub/docs/manual-ocaml/index.html)), functors are described in "Chapter 2 The module system" under "Section 3 Functors" while first-class modules are only mentioned in "Chapter 8 Language extensions" under "Section 6 First-class modules".

In my opinion, both first-class modules as well as "locally abstract types" (e.g. `fun (type a) (l : a list) -> ...`, cf Section 8.5) introduced in this PR are advanced features of OCaml that we should try to avoid in the public API unless it drastically simplifies the code.

As discussed, it would still be nice to *allow* the user to change the typing and/or reduction modules (and/or others) at run time. However I strongly believe that we should provide an interface allowing to do so (using first-class modules) while *also* allowing to ignore it and rely on functors instead.

Finally it's worth mentioning that functor instantiation is done at compile time, allowing code optimizations while first-class modules are values encapsulating modules in structures which does not allow such optimizations.

francoisthire commented:
> Note that in the official documentation of OCaml (caml.inria.fr/pub/docs/manual-ocaml/index.html), functors are described in "Chapter 2 The module system" under "Section 3 Functors" while first-class modules are only mentioned in "Chapter 8 Language extensions" under "Section 6 First-class modules".
> 

Note that there is a Chapter 6 with "Advanced examples with classes and modules" and that first-class modules are not mentioned while functor are.  I have the impression that from your point of view, I could conclude that functors are more "complex features" than first-class module. But of course, it would be unreasonable to jump to such conclusion. Is the order in the documentation compatible with the complexity of the feature of the language? Well, I think it is not and moreover, what we think as complex is really subjective. But let assume for a moment this is the case. In the API, we use a feature presented in section 8.15 called Extensible Variant Types. It is used in Basic.Debug, and hence is exposed in the API. Is that feature a so-called "advanced feature"? If so, does it simplifies the code drastically? We also tend to use features presented in section 8.2, 8.3, 8.4, 8.9, 8.10, 8.11, 8.13 in the API. 

Why can we assume that potential users of the API which currently are all members of the team knows these extensions, but not 8.6? I think we can't. I would like also to add an example which is objects and classes. For me, this is an advanced feature of OCaml which is presented in Chapter 7 and not in Chapter 8 as an extension of the language. Worst, it is even presented before modules and functors! (7.9 vs 7.10). Hence, is it easier to understand the classes of OCaml than any of the features presented in Chapter 8? Objects and classes are also presented in Chapter 3. But I tend to believe that you, me, or any other member of the team are not very familiar with object and classes of OCaml but are familiar with GADTs for example. 

This is lengthy, but I think you see my point, can we really base this debate between functors and first-class modules or their status as complex feature on the order these features are presented in the documentation?

> In my opinion, both first-class modules as well as "locally abstract types" (e.g. fun (type a) (l : a list) -> ..., cf Section 8.5) introduced in this PR are advanced features of OCaml that we should try to avoid in the public API unless it drastically simplifies the code.

I am sorry, but I disagree. What is called "advanced feature" is not universal, so except if you have a criteria which is better than the place where this feature is described in the documentation. 

Moreover, if you think that you cannot maintain the code because of first-class modules and locally abstract types, what we can do is to add my name as a maintainer of these files. This is what is done in Coq for example. 

> As discussed, it would still be nice to allow the user to change the typing and/or reduction modules (and/or others) at run time. However I strongly believe that we should provide an interface allowing to do so (using first-class modules) while also allowing to ignore it and rely on functors instead.

The old interface already allows to change the typing/reduction module at run-time thanks to first-class modules. However, the difference is that before, the user of the API has to use explicitly first-class module/functors in a way which i think is horrific. 

```ocaml
let foo (module R:Reduction.S) md entry =
    let (module E:Env.S) = (module Env.Make(R)) in
    ignore(Env.init md);
    let (module TC:Processor.S) = (module Processor.Typechecker(E)) in
    [...]
```
While we agree on the changes which remove the `ignore` line, the other lines are not really nice to write. This code has to be written everytime a function have a first-class module (R) in parameter which for my use cases happens several times. My point with this API was to change this behavior so that i can write 

```ocaml
let foo (module R:Reduction.S) env entry =
    let env' = Env.set_reduction_engine (module R) in
    [...]
```

Which is way better IMHO, and easier to understand. The problem is that I removed the functor `Env.Make` and `Processor.Typechecker`. Hence a potential user which had this code
```ocaml
module R = [...]
module E = Env.Make(R)
module TC = Processor.TypeChecker(E)

[...]
let foo md entry = TC.handle_entry md entry
```

may write

```ocaml
module R = [...]
let foo env entry = 
let env' = E.set_reduction_engine env (module R) in
TC.handle_entry env' entry
```

Sure, but there is another way to do it
```ocaml
module R = [...]
module Env = struct 
     include Env 
     
     let init file = 
          let env = init file in
          set_reduction_engine env (module R)

[...]
let foo env entry = TC.handle_entry env entry 
```

The point is that the module `Env` of the API is shadowed by this new `Env`.

Hence, at this stage, we can see that the new API and the old one allows to do the same thing in two different ways. Hence from there I see three choices:

1. Keeping the old API with what you call the good changes `Env.t`. 
2. Merge this PR
3. Enrich the API so that we get the best of the two worlds.

The point of this PR is to remove choice 1, I won't discuss it. The main issue with the point 3 are the motivations. No real user have the second usage for the moment, and even if it was the case, the solution I show is acceptable IMHO. But assume that this is not important, is it possible to keep the API as simple as possible and get the best of the two worlds? By as simple as possible, I mean an API such that there is one canonical choice to do one thing. In the case mentioned above, I mean if the user wants to change the rewrite engine, she does not have to choose between two different implementations such as functors and first-class modules. 

> As discussed, it would still be nice to allow the user to change the typing and/or reduction modules (and/or others) at run time. However I strongly believe that we should provide an interface allowing to do so (using first-class modules) while also allowing to ignore it and rely on functors instead.

This is a good point. But don't you think that the solution I showed earlier for the use-case number 2 solves this issue? The cost is one redirection everytime an `env` is created with `init`.

Gaspi commented:
I believe that (once API is made pure) this feature could be implemented in the current API simply using the functors to implement a new module of signature `S`:
```ocaml
module DynamicAPI :
sig
  include S
  val set_RE : (module S) -> t -> t
  val get_RE : t -> (module S)
end
  =
  struct
    type t =
      {
        sg : Signature.t;
        env: S
      }
    let init md = {Default.init md ; Default}
    let set_RE re e = { e with env = Make(re) }
    let get_RE e = e.sg
    (* Rest of the API *)
    let get_signature sg = sg.env.get_signature
    let get_symbols   sg = sg.env.get_symbols
    ...
    let unsafe_reduction sg = sg.env.unsafe_reduction
  end
```
This represents around 30 lines of codes in `Env.ml` (and 6 more in `Env.mli`), provides the module you need and still allows the user to statically use the functors if they don't need to change the reduction engine at runtime.

francoisthire commented:
The method set_re is not OCaml valid code.

On Sat, 21 Sep 2019, 08:09 Gaspard Férey, <notifications@github.com> wrote:

> I believe that (once API is made pure) this feature could be implemented
> in the current API simply using the functors to implement a new module of
> signature S:
>
> module DynamicAPI :sig
>   include S
>   val set_RE : (module S) -> t -> t
>   val get_RE : t -> (module S)end
>   =
>   struct
>     type t =
>       {
>         sg : Signature.t;
>         env: S
>       }
>     let init md = {Default.init md ; Default}
>     let set_RE re e = { e with env = Make(re) }
>     let get_RE e = e.sg
>     (* Rest of the API *)
>     let get_signature sg = sg.env.get_signature
>     let get_symbols   sg = sg.env.get_symbols
>     ...
>     let unsafe_reduction sg = sg.env.unsafe_reduction
>   end
>
> This represents around 30 lines of codes in Env.ml (and 6 more in Env.mli),
> provides the module you need and still allows the user to statically use
> the functors if they don't need to change the reduction engine at runtime.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Deducteam/Dedukti/pull/190?email_source=notifications&email_token=AATJVKYIM6ALJA55XYBYA43QKW3BDA5CNFSM4IXVTQK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7ILKZQ#issuecomment-533771622>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATJVK4KE7LTOFY4DECYIILQKW3BDANCNFSM4IXVTQKQ>
> .
>

francoisthire commented:
I think this is ready to be reviewed. Look the initial post to see the changes.
@GuillaumeGen @Gaspi @gabrielhdt @EmilieGrienenberger

gabrielhdt commented on kernel/basic.mli:
```diff
@@ -47,14 +47,20 @@ val dmark : ident
 
+module MidentSet : Set.S with type elt = mident
+
+module IdentSet : Set.S with type elt = mident
```
`elt = mident` or `elt = ident`?

gabrielhdt commented on kernel/basic.mli:
```diff
@@ -47,14 +47,20 @@ val dmark : ident
+module IdentSet : Set.S with type elt = mident
+
+module NameSet : Set.S with type elt = name
+
```
I already mentioned it, but it might be interesting to expose a comparison functions for each type, or modules like
```ocaml
struct
  type t = mident
  let compare = ...
end : OrderedType
```

gabrielhdt commented on kernel/rule.mli:
```diff
@@ -115,3 +117,7 @@ val pp_typed_context   : typed_context   printer
 
 val untyped_rule_of_rule_infos : rule_infos -> untyped_rule
+
+val check_arity : rule_infos -> unit
```
Some doc?

gabrielhdt commented on kernel/signature.mli:
```diff
@@ -18,33 +20,29 @@ type signature_error =
 
-val export              : t -> unit
-(** [export ()] saves the current environment in a [*.dko] file.*)
+val export              : t -> out_channel -> unit
```
Have you considered using formatters?

gabrielhdt commented on api/dep.mli:
```diff
@@ -1,42 +1,61 @@
-    deps: MDepSet.t; (** pairs of module and its associated path *)
-    ideps: ideps; (** up/down item dependencies *)
+    file     : string;            (** path associated to the module *)
+    deps     : Basic.MidentSet.t; (** pairs of module and its associated path *)
```
Documentation ought to be updated right? (pairs?)

gabrielhdt commented on api/dep.mli:
```diff
@@ -1,42 +1,61 @@
+
+(** {2 Dependencies function} *)
 
 val deps : t
```
Is it really necessary to use side effects?  Couldn't you have a function
```ocaml
find_deps : path list ->?(modules_only=true) -> ?(ignore=false) -> Basic.mident -> Entry.entry list -> data
```

gabrielhdt commented on api/dep.mli:
```diff
@@ -1,42 +1,61 @@
 val ignore : bool ref
+(** (default: [false]) If [true], no exception is raised if a [module] is not in the path *)
+
+val compute_all_deps : bool ref
```
Why not adding an optional argument to `make`?

gabrielhdt commented on commands/dkprune.ml:
```diff
@@ -0,0 +1,188 @@
+
+let _ =
+  let args = Arg.align
+    [ ( "-l"
```
`--verbose`?

francoisthire replied:
I am not sure about this one. Here I use `-l` as a shortcut to print `dkprune` logging output. While `verbose` would do more like printing logging from `dkprune` but `typing`, `reduction` etc...

gabrielhdt commented on api/env.mli:
```diff
@@ -36,85 +29,93 @@ val set_debug_mode : string -> unit
-  val get_type    : loc -> name -> term
-  (** [get_type l md id] returns the type of the constant [md.id]. *)
+val get_symbols : t -> Signature.rw_infos HName.t
+(** [get_symbols env] returns the content of the signature [sg]. *)
```
Could you describe what is bound to a key in the returned hashtable (I guess a symbol).

francoisthire replied:
It is obvous from the type right? I've added something though.

gabrielhdt commented on api/env.mli:
```diff
@@ -36,85 +29,93 @@ val set_debug_mode : string -> unit
 
-end
+val unsafe_reduction : t -> ?red:(Reduction.red_cfg) -> term -> term
+(** [unsafe_reduction env red te] reduces [te] according to the reduction configuration [red] *)
```
`unsafe` because it doesn't check whether the argument is well typed?  It would be sensible to say why it is unsafe in the documentation.

francoisthire replied:
Good point. I've added a comment to say that the term was not type checked.

rlepigre commented on kernel/signature.mli:
```diff
@@ -84,7 +84,9 @@ type rw_infos =
     (** The type of a symbol *)
     rules         : rule_infos list;
-    (** The list of rules associated to a symbol. They are ordored by their declaration within a file and in order they are imported in the signature *)
+    (** The lifo pile of rules associated to a symbol.
```
What is this "lifo pile" terminology here?
This data structure is called a stack in English.
And a the "fifo" version is called a queue.
